includes/admin/notes/class-wc-notes-run-db-update.php

```diff
@@ -0,0 +1,312 @@
+	 *
+	 * @param int $note_id Note id to update.
+	 */
+	private static function update_in_progress_notice( $note_id ) {
+		// Same actions as in includes/admin/views/html-notice-updating.php.
+		$pending_actions_url = admin_url( 'admin.php?page=wc-status&tab=action-scheduler&s=woocommerce_run_update&status=pending' );
+		$cron_disabled       = Constants::is_true( 'DISABLE_WP_CRON' );
+		$cron_cta            = $cron_disabled ? __( 'You can manually run queued updates here.', 'woocommerce' ) : __( 'View progress →', 'woocommerce' );
+
+		$note = new WC_Admin_Note( $note_id );
+		$note->set_title( __( 'WooCommerce database update in progress', 'woocommerce' ) );
+		$note->set_content( __( 'WooCommerce is updating the database in the background. The database update process may take a little while, so please be patient.', 'woocommerce' ) );
+
+		$note->clear_actions();
+		$note->add_action(
+			'update-db_see-progress',
+			$cron_cta,
+			$pending_actions_url,
+			'unactioned',
+			false
+		);
+
+		$note->save();
+	}
+
+	/**
+	 * Update the existing note with $note_id with information that db upgrade is done.
+	 *
+	 * This is the last notice (3 out of 3 notices) displayed to the user.
+	 *
+	 * @param int $note_id Note id to update.
+	 */
+	private static function update_done_notice( $note_id ) {
+		$hide_notices_url = html_entity_decode( // to convert &amp;s to normal &, otherwise produces invalid link.
+			wp_nonce_url(
+				add_query_arg(
+					'wc-hide-notice',
+					'update',
+					admin_url( 'admin.php?page=wc-settings' )
+				),
+				'woocommerce_hide_notices_nonce',
+				'_wc_notice_nonce'
+			)
+		);
+
+		$note = new WC_Admin_Note( $note_id );
+		$note->set_title( __( 'WooCommerce database update done', 'woocommerce' ) );
+		$note->set_content( __( 'WooCommerce database update complete. Thank you for updating to the latest version!', 'woocommerce' ) );
+
+		$actions = $note->get_actions();
+		if ( ! in_array( 'update-db_done', wp_list_pluck( $actions, 'name' ) ) ) {
+			$note->clear_actions();
+			$note->add_action(
+				'update-db_done',
+				__( 'Thanks!', 'woocommerce' ),
+				$hide_notices_url,
+				'actioned',
+				true
+			);
+
+			$note->save();
+		}
+	}
+
+	/**
+	 * Return true if db update notice should be shown, false otherwise.
+	 *
+	 * If the db needs an update, the notice should be always shown.
+	 * If the db does not need an update, but the notice has *not* been actioned (i.e. after the db update, when
+	 * store owner hasn't acknowledged the successful db update), still show the notice.
+	 * If the db does not need an update, and the notice has been actioned, then notice should *not* be shown.
+	 * The same is true if the db does not need an update and the notice does not exist.
+	 *
+	 * @return bool
+	 */
+	private static function should_show_notice() {
+		if ( ! \WC_Install::needs_db_update() ) {
+			try {
+				$data_store = \WC_Data_Store::load( 'admin-note' );
```

Comment:
note that this needs wc-admin tables to be installed. so we may not see notices immediately after activation if table generation script is still running or (maybe) if plugin is updated via ftp instead of wp-admin. have we tested this scenario?

Reply:
Good point, checking.

Reply:
Looks like intial db install happens synchronously on first request after updation, so I guess we should be here. I tested out by replacing WC 3.9 without wc-admin with WC 4.0 via SFT and faced no issues. Resolving this conversation.